authorization/v2/uniformResourceIdentifier.go

```diff
@@ -24,6 +26,79 @@ func NewUriFromRequest(resource resources.Resource, handler gin.HandlerFunc, ctx
 	}
 }
 
+// NewURIFromString parses the string representation of a URI into the UniformResourceIdentifier struct.
+// NewURIFromString expects the string to be of the following form, otherwise ErrInvalidURI is returned.
+// hs:<service_name>:<subsystem>:<version>:<category>:<resource_name>?<allowed_arguments>#<permission_metadata>
+func NewURIFromString(source string) (UniformResourceIdentifier, error) {
+	remainingURI, metadata, err := newURIListFromString(source, "#")
+	if err != nil {
+		return UniformResourceIdentifier{}, errors.Wrap(ErrInvalidURI, err.Error())
+	}
+
+	remainingURI, arguments, err := newURIListFromString(remainingURI, "?")
+	if err != nil {
+		return UniformResourceIdentifier{}, errors.Wrap(ErrInvalidURI, err.Error())
+	}
+
+	return UniformResourceIdentifier{
+		path:      remainingURI,
+		arguments: arguments,
+		metadata:  metadata,
+	}, nil
+}
+
+func newURIListFromString(source string, sep string) (remainingURI string, uriList map[string]string, error error) {
```

Comment:
I understand that the name of the function is a bit confusing based on what the function actually does. If it is renamed to `extractURIListFromString` then I think it would represent it better. I personally don't see returning the `remainingURI` as too hacky, especially since it's a private helper function for `NewURIFromString`. In fact, this function was only introduced since the code was duplicated in `NewURIFromString` to extract the metadata and arguments.

I tried a few different methods, but they didn't feel that clean and mostly just meant you had to check for more edge cases. As an example, if you try getting the index of each rune (`#` or `?`) and use array slices to extract the metadata/args, it means you have to perform checks every time to ensure you don't go outside the bounds of the array and (to me anyway) seemed harder to read than this method.
Additionally, I think using this allows for more strict parsing since you can explicitly check that the URI only contains a single `?` and `#`.

In terms of convention where you shouldn't return more than two variables, that's something that could be avoided by passing the uriList by a reference parameter.

I'm very happy to hear ideas for alternative methods and I implement them, but to me, this method is quite succinct and readable.

Reply:
Ok let's just rename it to `extractURIListFromString ` and go with that. It's a really minor thing anyway